### PR TITLE
FEATURE: Add configurable actions on node creation

### PR DESCRIPTION
**Pull request by dimaip**
This feature allows to run arbitrary node transformations on the node that is being created. This can be useful in a number of cases, e.g. you can pre-populate an article node with a text node inside main content collection. Unlike `childNodes`, it doesn't make these auto-created nodes rigid, and allows you to alter or remove these nodes later on.

Here is an example configuration, that sets some prefix to the page title when it is created::

    'TYPO3.Neos.NodeTypes:Page':
      options:
        actions:
          onCreate:
            -
              type: 'ChangePropertyValue'
              settings:
                property: 'title'
                newValue: '${q(node).property("title") + " - this suffix is set in an action"}'

NEOS-1166 #comment This feature is the basis for the wizard functionality, but can be merged and used stand-alone.

### CODE REVIEW

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+use TYPO3\TYPO3CR\Domain\Service\NodeTypeManager;
+
+/**
+ * Functions for executing actions on node creation (e.g. creating child nodes or setting properties based on wizard
+ * data)
+ *
+ * @Flow\Scope("singleton")
+ */
+class NodeActionsService
```
Comment (dimaip) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: I don't like the name. Maybe name it smth like `NodeTransformationActionsSerive` or `OnNodeEventActions`?

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+        return $contextVariables;
+    }
+
+    /**
+     * @param array $settings
+     * @param array $contextVariables
+     * @return array
+     */
+    protected function parseSettings(array $settings, array $contextVariables)
```
Comment (dimaip) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: I would like to move eel parsing to `NodeTransormation` service, it would be quite useful for normal node transformations. But the problem is that NodeTransformations operate on NodeData, and not on Node, so it would be hard to build an EEL context from them. Or is it okay to just build a Node from NodeData + live context?

File: TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst [hunk 3]
```diff
@@ -250,6 +250,34 @@ The following options are allowed:
...
+
+    ``onCreate``
+      Array of actions that get triggered when the node is being created. Each action represents a node transformation that may have a type and a set of settings.
+
+      ``type``
+        A type of the transformation. TYPO3CR package provides a set of default transformations: ``AddDimesnions``, ``AddNewProperty``, ``ChangeNodeType``, ``ChangePropertyValue``, ``RemoveNode``, ``RemoveProperty``, ``RenameDimension``, ``RenameNode``, ``RenameProperty``, ``SetDimensions`` and ``StripTagsOnProperty``.
+        But it is possible to define a custom transformation by providing a class that implements a ``TransformationInterface``.
+
+      ``settings``
+        A set of setting for given transformation. Consult the source code to see which transformation accepts what settings. Each setting key is parsed with EEL, and has the ``node`` context variable set to current created node.
```
Comment (daniellienert) [inline TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst]: setting for *the* given transformation

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php (hunk 1 above)
Comment (daniellienert) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: I would prefer `NodeTransformationActionsSerive` as it is about actions on NodeTransformation not actions of a node.

**Comment by daniellienert** [comment]
Hey @dimaip - this is a very nice feature! - we already use the wizzard feature from the Foundation package. Would it be possible to have a functional test for the service?

**Comment by dimaip** [comment]
@daniellienert can you guide me on what this functional test test exactly? I suck at TDD

**Comment by theilm** [comment]
Great work, @dimaip! I like the idea to use transformations. We limit the use case of the actions to transformations (you can't or should not use transformations to call a web service wh
Great work, @dimaip! I like the idea to use transformations. We limit the use case of the actions to transformations (you can't or should not use transformations to call a web service when creating a new blog post for example) - but that's probably a good thing. 

The main drawback is indeed the node data handling. In Neos services, we already operate with nodes; then we pass the node data to the transformation; and there, it gets converted to a node again. Internally, this would be not a big deal probably. But as we stick to the TransformationInterface, we encourage users to do it the same way.

As discussed via Slack, we should probably try to change the transformation interface to use Node instead of NodeData.

Concerning the service name. I would suggest to change "actions" to "transformations" everywhere. Replace "actions" with "transformations" in configuration and call the service "NodeTransformationService".

**Comment by dimaip** [comment]
@theilm I had the same idea about renaming it to transformations. @kdambekalns had an argument, that from user POV it doesn't matter what lies underneath the actions engine, that for user it shouldn't matter that much.
But I somehow still feel that renaming it would actually cause less confusion. In the first place, our integrators are not idiots, and do know what transformations are, so honestly saying that this is just that makes sense.
@kdambekalns wdyt?

Regarding doing other non-transformation stuff, there are signal for that. In theory you could provide your own fake transformation that would call some service, but that would be ugly, and using signals for that makes much more sense.

**Comment by dimaip** [comment]
Just spoke with @kitsunet, he's into changing NodeData to Node on next breaking release of Neos. For now we can leave it as is.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php [hunk 4]
```diff
@@ -32,14 +35,49 @@ class NodeTransformation
...
+     * @var array
+     */
+    protected $contextVariables;
+
+    /**
+     * @return void
+     */
+    public function initializeObject()
+    {
+        if ($this->eelEvaluator instanceof \TYPO3\Flow\Object\DependencyInjection\DependencyProxy) {
```
Comment (kdambekalns) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php]: Why is this needed?

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Transformations/CreateNodes.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    /**
+     * If the given node has the property this transformation should work on, this
+     * returns TRUE.
+     *
+     * @param NodeData $node
+     * @return boolean
+     */
+    public function isTransformable(NodeData $node)
+    {
+        return ($node instanceof NodeData);
```
Comment (kdambekalns) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Transformations/CreateNodes.php]: If the method can be called, it must be a NodeData instance (type hint, eh?), so no need to check anything.

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
+<?php
+namespace TYPO3\Neos\Service;
+
+/*                                                                        *
+ * This script belongs to the TYPO3 Flow package "TYPO3.Neos".            *
```
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: Old header format…

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php (hunk 1 above)
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: A related thought: this is a domain service, not a REST service, so it should be moved as well.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php (hunk 4 above)
Comment (theilm) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php]: $this->eelEvaluator is used as a parameter in the evaluateEelExpression method which expects a EelEvaluatorInterface. If you don't make sure the proxy is activated, at the time you pass the object to the function it could still be a DependencyProxy object.
Instead of activating the dependency in the initializeObject method, we could deactivate lazy injection via annotation. Or we could do the proxy activation like now but right before the evalualeEelExpression call to prevent unneeded activation in transformations that don't use EEL. wdyt?

**Comment by kdambekalns** [comment]
Fixed some nitpicks. I agree to the service weird name that has to change, will do that (and the move) later. Also I am working on documentation, I'll cough up some generic migrations docs while we are at it… So, stay tuned.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php (hunk 4 above)
Comment (kdambekalns) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Migration/Service/NodeTransformation.php]: Thanks for the explanation. Indeed switching off lazy injection should be done. I'll update the code…

**Comment by kdambekalns** [comment]
Ok, pushed documentation and moved the service. The name is not changed yet.

File: TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php (hunk 1 above)
Comment (kdambekalns) [inline TYPO3.Neos/Classes/TYPO3/Neos/Service/NodeActionsService.php]: It is not about actions happening on transformations, it is about actions happening on nodes (or on node lifecycle events). So I left the name unchanged in my last commit.

**Comment by kdambekalns** [comment]
Regarding the naming, I still think "actions" is fine. From a users perspective actions are triggered on lifecycle events, that these (currently) use transformations (only) to do their work, should not trick us into naming this "technically correct". Later more actions could be added, things not using transformations, and then there would be even more confusion…

Another thing: would it make sense to allow to have multiple dynamic properties on the new CreateNodes transformation? Could be done by removing `dynamicProperty` and `dynamicPropertySource` and just have `dynamicProperties` being an array of property names and values… I could imagine this will be needed sooner or later. And adding support for that later would be a breaking change and/or lead to a really ugly syntax.

**Comment by dimaip** [comment]
`dynamicProperties` - :+1: 

> From a users perspective actions are

Well, it's not exactly users who edit nodetypes or add custom transformations... it's the integrator. And this integrator must already know how migrations/transformations work, or else he would not be able to use the feature.
If we add functionally in the future that would not be related to transformations, it would just get a different name. How do you think to change this functionality in B/C way to stop using Transformations in the future...

**Comment by kdambekalns** [comment]
Well, integrators are the users of that "API" - they are our clients, they are the users.

About adding new ways of doing actions… The service we have could just fetch an object of the given type, and then if it's a transformation, call it as needed, if it's something else, do something else. Entirely internal changes, the "API" in the settings could stay the same.

**Comment by dimaip** [comment]
Read through your commits. Crazy work!

My argument was that for immediate future, the "transformations" name would be more telling to integrators than the "actions", because it already implies without looking at the docs that this thing uses transformations. But if we have such far-fetching plans regarding the future development of this API, then heck why not, I don't really mind.

**Comment by kdambekalns** [comment]
Well, I don't know about out far-fetched plans, but it is about "some action" happening on node creation (update, removal, …). :)

**Comment by theilm** [comment]
Just re-thought about actions, transformations etc. I am fine with the actions naming. But I guess we have to clarify the difference between the signal that is emitted on node creation and the new actions. The actions are not called every time a node is created, but only if it's created via the Rest service (which means the UI in most cases). Perhaps we could omit the term "lifecycle" in the documentation to prevent confusion? If you agree, I'll try to adapt documentation accordingly.

**Comment by dimaip** [comment]
@theilm Yes, that should be clarified.

**Comment by robertlemke** [comment]
Hey @dimaip – looks like a really good feature, thank you very much for implementing it! It went a bit under my radar, so I'm a bit late with my feedback and I'm sorry if that might cause additional work on your side. Either way, I took two hours this morning to think about the feature and also discussed it thoroughly with Christian and Karsten who happened to be around.

If you don't mind, I would like to give some feedback from a conceptual / user's point of view. I haven't tried out the feature so far and didn't examine the code in detail, but I have a few comments and questions after reading the documentation and comments.

#### When actions are triggered

In the documentation you mention:

> Please note: Node actions are only executed when working with the Neos backend (or the Neos specific NodeOperations service). If you need to trigger general lifecycle actions that are not bound to editing, you can use the signals  emitted by the Node class.

As a developer / user I would expect the actions to be consistent and be triggered whenever the even occurs. So, every time a node was created, I would expect that also the respective action would be executed. @theilm mentioned, the difference between the signals and these actions should be clarified; I'd go further and would like to ask if there's a conceptual reason for making this difference, as it can be quite confusing to site integrators and developers.

I suggest that the action triggers directly correspond to (a selection of) signals in the `Node` Domain Model. I know that the signals there are a bit inconsistent at the moment (some are named after the method they appear in and others rather describe the event which led to emitting the signal). So let's be careful with our selection of signals we want to support in the initial implementation, as we implicitly make them part of the public API.

If we take a peek into the future, we will likely refactor the Content Repository to an even-based model, likely based on Event Sourcing and possibly also CQRS. So, if we take some of the signals today which use past tense in English (`nodeAdded`, `nodeUpdated`) we are relatively safe for when we refactor them to Domain Events. Let's not use the other signals yet which don't follow that naming scheme (`afterNodeCreate`, `BeforeNodePropertyChange` and so on).

So, practically, create some dispatcher class which listens to the respective signals and then calls the action implementations as necessary.

#### Options

Regarding the "options" section in the Node Type configuration: in the documentation you write "Miscellaneous configuration options" which hints at a bit of a weak naming because this will become the section for "all the rest". @kdambekalns explained that this section actually existed already in TYPO3CR - but I suggest that we don't use it, but rather move "actions" to the top level of the Node Type Configuration and adjust the YAML schema accordingly.

#### Transformations

I learned that @kdambekalns originally suggested to re-use transformations in order to avoid code duplication. After discussing this a bit, I think that it's not such a good idea anymore, for three reasons:

1. Transformations are based on Node Data, and we must keep developers and especially site integrators from using Node Data directly wherever possible. Also, when we work at such a low level, signals are not triggered, and other code might rely on that information to keep data consistent.

2. By using the Transformation Interface, we would implicitly make all transformations part of the public API (even if we state the opposite). I think that they are not ready to be public API, especially because they work with an internal API of the Content Repository

3. Most of the existing transformations are not suited for being used in actions, for example changing the content dimension. We risk that people will fall into the trap of just using any of them.

So, let's revoke that decision, if possible. I suggest that you create a lean `NodeActionInterface` which simply supports some execute method and a way to pass action options (or "arguments", which would be my preferred name for it).

Let's implement only a few Node Actions in the beginning, but allow integrators to specify a fully qualified class name to their own custom Node Actions.

##### Bottom Line

I hope that my comments are helpful for making this a polished new feature. Let's not mention "Transformations" anywhere related to this feature, let's establish the term "Node Actions" (or Node Events Handlers?) and provide a consistent API and naming wherever possible.

If you need any help or have questions, I'll try to respond in a more timely manner ;)

**Comment by theilm** [comment]
Hey @robertlemke thanks for the comments.

#### When actions are triggered

I think we should stick to the conecpt. The actions should be really related to editing. Think of it as some kind of boilerplate functionality: You create a tabs element, some tabs are created automatically to make life easier for the editor. If you create a tab node via PHP for example, you probably don't want dummy / boilerplate sub nodes to be created. Same goes for changing node properties.

I used signals to create sub nodes in a project to make life for editors easier - and it was really annoying that the nodes were created also when doing automated stuff in command controllers.

Another reason why we integrated it in the REST service that it used in the backend and we don't use the Node signals: We want to add a wizard functionality to be able to add additional properties. Think of it as a form on node creation; the properties of the form are made available in the EEL context and can be used in the actions. Simple example: Set the document title in a dialog, not in the tree. @dimaip already did some stuff for the wizard: https://github.com/neos/neos-development-collection/pull/27
The additional (optional) argument needed to support this in the NodeService is already part of this PR.

#### Transformations

I like the idea to use transformations, but I understand your concerns. We already discussed the NodeData problems - but as @kitsunet confirmed to @dimaip that we could probably change the TransformationInterface to use Node instead of NodeData in the next breaking release, we decided to go for it. But it's possible to go back to the initial idea and make a new interface. @dimaip wdyt?

**Comment by robertlemke** [comment]
@theilm regarding signals: what I meant was to subscribe to signals internally in order to trigger actions. So I'm all in favour of introducing the Actions concept, but I strongly suggest to bind them to the existing signals internally, because otherwise we'll get into weird situations when actions are not triggered when you would expect them to be triggered.

As for the transformations: I don't think that its technically feasible to refactor them to `Node` instead of `NodeData`, because they need that low level access to data in order to commit certain actions (for example, fixing node structures which currently wouldn't be allowed).

**Comment by dimaip** [comment]
Hey Robert! Thanks a lot for your thoughtful feedback!

> As for the transformations: I don't think that its technically feasible to refactor them to Node instead of NodeData, because they need that low level access to data in order to commit certain actions (for example, fixing node structures which currently wouldn't be allowed).

It's quite easy to get the `NodeData` from `Node`, I don't see how switching to `Node` would limit migrations capabilities in any way. But migrations are publicate interface too, and it's a bit awkward that we expose internal stuff there.
I don't really mind switching back to our own interface, we just have to agree on something here. I switched to transformations based on recommendation from @kdambekalns, and I still like that idea.

I completely support what @theilm said about signals. And I don't understand how would you pass data from the wizard to the signal?

**Comment by kdambekalns** [comment]
While you can get `NodeData` from a node, you might need to have transformations applied to `NodeData` that cannot be hydrated as a node (making the transformation necessary in the first place). So I second that thought, don't think `Node` can be used in transformations.

**Comment by dimaip** [comment]
@kdambekalns makes sense. So what do we decide on:

- Switch back to own interface and move back to Neos package?
- I would leave eel parsing of transformations anyways, it's just cool, maybe in a separate PR
- Move yaml configuration a level up. We have to agree on a name here. Like `actions.nodeAdded`?
- I strongly insist on leaving those actions tied with REST for the reasons mentioned be me and @theilm above

**Comment by dimaip** [comment]
@robertlemke ah and one more thing, it was @theilm who implemented this feature, not me of course ;)
I'm taking care of the wizard UI part.

**Comment by robertlemke** [comment]
@dimaip the problem is that sometimes you can't really instantiate a valid Node with a proper context and so on, because there will be situations where such a Node wouldn't be valid (anymore) and the migration would be there to solve exactly that inconsistent state. So, we need to provide that low level interface to transformations and I really wouldn't re-use them for the purpose you have in mind.

Side note: in my point of view, Node Migrations haven't been (yet) part of the public API, but of course one can argue about it.

Regarding the signals: I think we may still be talking cross-purposes here:

If you look at the feature this particular change is about, what you learn from the documentation is, that you can configure actions for a node type which is triggered in certain situations, for example after a node has been created. So, if the intention is to make this a generic mechanism which is executed automatically, it's important that actions are executed whenever the respective event happens, no matter how you achieve that technically.

Now, if you create a wizard which wants to deliver additional information, you somewhat bypass the interface which developers have in the Node Types configuration. If the intention for the wizard is that you have some wizard which collects information and then executes an action parameterized by this collected information, you might not need a generic action feature in Node Types, but rather some way to configure actions, given them a name and then just use that in your wizard (for example, "when this wizard finished, then call action 'create-new-press-release-action' and see details for that in some corresponding settings).

Before we agree or disagree on that - do you understand what I mean with that?

**Comment by dimaip** [comment]
How I feel about these things, is that actions are an integrator/UI facing feature. It's main intent is to generate some template content, that differs from the permanent nature of `childNodes` in a way that you can later delete or move it. Actions may take data from the UI wizard to batch-create more stuff, e.g. create a grid with X rows, upload an array of images and create an Image node for each image etc.

For more lowlevel stuff that should not be tied to UI, a more low-level concept of signals may already be used, and I don't think we should provide a configuration for that from the nodetypes configuration.
If you think we need both things, we could have two types of actions in configuration, e.g. `nodeAdded` for low-level signal stuff and `nodeCreatedFromUI` (think of a better name) for the REST stuff.

I don't think I quite understood your last paragraph correctly though. If you mean to submit another ajax request that would call a different endpoint, in addition to the REST request when the node is created, I don't think that's a cool idea, there will be race-conditions between two requests etc.

**Comment by dimaip** [comment]
So to put it into better words, we have two clearly different intents here:

1. Stuff that should only happen on user UI actions (templating etc). The result of such actions is not tightly coupled to the underlying node structure, it's just a quick way to batch-create content.
2. Low-level stuff that should happen on any Node manipulations. Right now we signals for that, and it maybe a good idea to create a more declarative ways to configure them as well. But these are two very different things. I think a clear solution would be to have two action types for those.

**Comment by robertlemke** [comment]
@dimaip okay, I think it's getting clearer for me now.

As far as I understood you, you are at the moment rather looking for a way to configure (a set of) actions which can be referred to and executed by the wizard. You are not looking for a solution for calling actions without the wizard being involved (for example, automatically whenever a property a of a specific content type was updated).

If that's the case, what's mostly misleading, in my opinion, that you configure these actions in the Node Types schema. As I wrote earlier, I think that gives the impression that these action would be executed automatically, for example also when someone creates a custom command controller importing content from a shop. I think that this would be also a very valid use case for such a feature - for example, when importing information about hotels from a booking system and creating `Hotel` document nodes, your new feature could automatically create a whole sub structure (Pages, `Room` Nodes and so on).

I'd say, if it's not fully automatic and happening every time (a node is created, for example), then it must not be configured in NodeTypes.yaml but be some `actions` section in Neos' Settings.yaml.

**Comment by dimaip** [comment]
@robertlemke I think I still don't fully get what I mean here.

> As far as I understood you, you are at the moment rather looking for a way to configure (a set of) actions which can be referred to and executed by the wizard.

No, absolutely not. Let's take a basic example. An Article nodetype with `main` content collection. When the user creates an Article, I would like to have a Text node types automatically created for him in the main content collection, I hope you understand just by how much that would improve the UX. This example has nothing to do with the wizard. Wizard is just a particular case, where you may get some parameters, based on which you want these actions to be customized.

**Comment by dimaip** [comment]
> As far as I understood you, you are at the moment rather looking for a way to configure (a set of) actions which can be referred to and executed by the wizard.

Yet again, the wizard is not just some wizard that has some place in Neos UI independant from node types. It's the wizard on node creation, its input elements are also defined in NodeTypes.yaml, it is semantically tied to given nodetype, it's an indispensable part of it, actually. Will provide more examples.

**Comment by dimaip** [comment]
This is probably the most meaningful example to illustrate my point:

```
'TYPO3.Neos.NodeTypes:Gallery':
  ui:
    label: 'Gallery'
    wizard:
      properties:
        assets:
          type: array<TYPO3\Media\Domain\Model\Asset>
          ui:
            label: 'Assets'
  options:
    actionsOnNodeCreation:
      createImages:
        type: 'createNodesFromArray'
        options:
          dynamicPropertySource: '${data.assets}'
          dynamicProperty: 'image'
          nodeData:
            'nodeType': 'TYPO3.Neos.NodeTypes:Image'
```

Don't mention it, this example uses the old configuration syntax, but the point should be quite clear: both the wizard and the action creating nodes with data from the wizard are closely related to the Gallery node type. And again, when creating Wizard nodes programmatically, I would not want any nodes to be implicitly created, I would create Image child nodes myself.
E.g. in typo3 we had some wizards to batch-create nodes (can't remember the name), now this stuff is NOT it. It's an integral property of nodetype's UI, but not of its data model.

**Comment by dimaip** [comment]
Yet again, a valid (and useful) compromise between these two views would be to create to action types: one that is triggered only form UI(REST), the other on signals.

**Comment by dimaip** [comment]
@robertlemke  Btw, if we want `actions` to be top level in the schema, and not within `options`, then we would have to define it in TYPO3CR and not in Neos package, and that's not cool.

**Comment by dimaip** [comment]
Me and @theilm gave it a bit more thought. I think we've come up with something cool, declarative and userlandish. Will post later tonight, hope you'll love.
@robertlemke thanks for not letting us merge lame crap!

**Comment by dimaip** [comment]
Closing it down in favour of https://discuss.neos.io/t/rfc-node-templates/715
Thanks @robertlemke and sorry for your wasted time reviewing it, but I think the idea with which we came up eventually thanks to you is actually much better!

**Comment by robertlemke** [comment]
@dimaip thanks a lot for considering my feedback and for not being discouraged ;-)

**Comment by kdambekalns** [comment]
See https://github.com/neos/neos-development-collection/pull/268 for the extracted migrations documentation.
